config.org
:PROPERTIES:
#+LATEX_COMPILER: xelatex
#+LATEX_CLASS: elegentpaper
#+OPTIONS: prop:t
:END:
#+title: Doom Emacs literal configuration
#+author: chuj

* Table of Contents :TOC:noexport:
- [[#misc][Misc]]
  - [[#lexical-binding][Lexical-binding]]
  - [[#user-info][User info]]
  - [[#keyfreq-blacklist][Keyfreq blacklist]]
  - [[#not-tab][Not tab]]
  - [[#encoding][Encoding]]
  - [[#load-path][Load path]]
- [[#theme][Theme]]
  - [[#main-theme][Main Theme]]
  - [[#font][Font]]
  - [[#line-number][Line Number]]
  - [[#ruler][Ruler]]
  - [[#ident-guide][Ident guide]]
  - [[#custom-set-face][Custom set face]]
  - [[#full-screen][Full screen]]
  - [[#mode-line][Mode line]]
- [[#global-enable-mode][Global enable mode]]
- [[#programming][Programming]]
  - [[#prog-mode-hook][prog-mode hook]]
  - [[#syntax-highlight][Syntax highlight]]
  - [[#lsp][Lsp]]
  - [[#company-mode][company mode]]
  - [[#cc][C/C++]]
  - [[#rust][Rust]]
  - [[#common-lisp][Common Lisp]]
- [[#org-mode][Org-mode]]
  - [[#general][General]]
  - [[#storage][Storage]]
  - [[#latex][Latex]]
  - [[#for-export][For export]]
- [[#key-bindings][Key bindings]]
- [[#other][Other]]
  - [[#wsl-setup--chinese-input-rime][WSL setup && Chinese input (rime)]]
  - [[#profile-startup][Profile startup]]
  - [[#eaf][EAF]]

* Misc
** Lexical-binding
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+end_src
** User info
Some functionality uses this to identify you, e.g. GOG configuration, email clients, file templates and snippets. It is optional.
#+begin_src elisp
(setq user-full-name "chuj"
      user-mail-address "1768485949@qq.com")
#+end_src
** Keyfreq blacklist
#+begin_src emacs-lisp
(setq keyfreq-excluded-commands
      '(self-insert-command
        forward-char
        backward-char
        previous-line
        next-line
        evil-previous-line
        evil-next-line
        evil-backward-char
        evil-forward-char
        evil-force-normal-state
        evil-insert
        term-send-raw
        org-self-insert-command
        vterm--self-insert
        evil-normal-state
        exit-minibuffer
        vertico-next ; vertico candidate chose
        mwheel-scroll
        ))
#+end_src
** Not tab
don't use tab
#+begin_src emacs-lisp
(setq-default indent-tabs-mode nil)
(setq-default tab-width 2)
#+end_src
** Encoding
让所有的编码都变成 UTF-8
#+begin_src emacs-lisp
(set-language-environment               "UTF-8")     ;; System default coding
(prefer-coding-system                   'utf-8)      ;; prefer
(set-buffer-file-coding-system          'utf-8-unix) ;;
(set-charset-priority                   'unicode)    ;;
(set-clipboard-coding-system            'utf-8-unix) ;; clipboard
(set-default-coding-systems             'utf-8)      ;; buffer/file: 打开文件时的默认编码
(set-file-name-coding-system            'utf-8-unix) ;; unix/linux/macos
(set-keyboard-coding-system             'utf-8-unix) ;; keyboard
(set-next-selection-coding-system       'utf-8-unix) ;; selection
(set-selection-coding-system            'utf-8)      ;; selection
(set-terminal-coding-system             'utf-8-unix) ;; terminal
(setq coding-system-for-read            'utf-8)      ;;
(setq default-buffer-file-coding-system 'utf-8)      ;;
(setq locale-coding-system              'utf-8)      ;; local
#+end_src
** Load path
#+begin_src emacs-lisp
(add-to-list 'load-path "~/.doom.d/lisp")
#+end_src
* Theme
** Main Theme
use doom-dark + theme (a vs-code like dark theme), and use a blue modeline
+stick to doom one+
#+begin_src emacs-lisp
(setq doom-theme 'doom-dark+
      doom-dark+-blue-modeline t)
#+end_src
** Font
使用 JetBrains Mono 字体作为等宽字体

非等宽字体使用 Myriad Pro

unicode 字体使用 /我在这里下载的[[https://github.com/samuelngs/apple-emoji-linux][Apple Color Emoji]]/
#+begin_src emacs-lisp
;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 20
                           :weight 'normal)
      doom-variable-pitch-font (font-spec :family "Myriad Pro" :size 26)
      doom-unicode-font (font-spec :family "Apple Color Emoji" :size 20))

;(set-fontset-font "fontset-default" 'han '("LiHeiPro" . "ttf"))
#+end_src
** Line Number
This determines the style of line numbers in effect. If set to `nil', line numbers are disabled. For relative line numbers, set this to `relative'.

#+begin_src emacs-lisp
(setq display-line-numbers-type 'relative)
#+end_src
** Ruler
a 80 ruler
#+begin_src emacs-lisp
(setq display-fill-column-indicator-column 80)
;; we don't want a ruler in shell right? so only in prog mode
(add-hook 'prog-mode-hook #'display-fill-column-indicator-mode)
#+end_src
** Ident guide
#+begin_src emacs-lisp
;; better indent guide
(add-hook 'prog-mode-hook 'highlight-indent-guides-mode)
(setq-default highlight-indent-guides-method 'character)
#+end_src
** Custom set face
#+begin_src emacs-lisp
;;; custom set face
(custom-set-faces
;  '(rainbow-delimiters-depth-1-face
;    ((t (:inherit rainbow-delimiters-base-face :foreground "gold"))))
;  '(rainbow-delimiters-depth-2-face
;    ((t (:inherit rainbow-delimiters-base-face :foreground "LightGreen"))))
;  '(rainbow-delimiters-depth-3-face
;    ((t (:inherit rainbow-delimiters-base-face :foreground "magenta1"))))
  '(rainbow-delimiters-depth-3-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "LightGreen"))))
  '(rainbow-delimiters-depth-4-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "gold"))))
  '(lsp-face-semhl-interface
    ((t (:inherit font-lock-variable-name-face))))
;  '(rainbow-delimiters-depth-4-face
;    ((t (:inherit rainbow-delimiters-base-face :foreground "beige"))))
  '(mode-line ((t (:family "Noto Sans" :height 0.9))))
  '(mode-line-active ((t (:family "Noto Sans" :height 0.9)))) ; For 29+
  '(mode-line-inactive ((t (:family "Noto Sans" :height 0.9))))
  ;; i decide to not enable whitespace mode by default
  ; set tab to the background color
  '(whitespace-tab ((t (:background "1e1e1e"))))
; (face-attribute 'default :background) => 1e1e1e
  )
#+end_src
** Full screen
#+begin_src emacs-lisp
(add-to-list 'default-frame-alist '(fullscreen . maximized))
#+end_src
** Mode line
doom modeline configuration （看起来 doom 的 modeline 还是有些卡卡的，所以暂时不用了）
#+begin_src elisp
(setq doom-modeline-highlight-modified-buffer-name t)
(setq doom-modeline-major-mode-color-icon t)
(setq doom-modeline-indent-info t)
;; used for modeline's font, see custom set face
(setq doom-modeline-height 1)
#+end_src

custom modeline configuration
*** line + column numbers display
#+begin_src emacs-lisp
(column-number-mode 1)
#+end_src
*** simplify the cursor position
使用百分比 + top + bottom 这样来显示所处的位置对我来说多少有些不必要，所以取消之
#+begin_src emacs-lisp
(setq mode-line-position
      '(;; %p print percent of buffer above top of window, o Top, Bot or All
        ;; (-3 "%p")
        ;; %I print the size of the buffer, with kmG etc
        ;; (size-indication-mode ("/" (-4 "%I")))
        ;; " "
        ;; %l print the current line number
        ;; %c print the current column
        (line-number-mode ("%l" (column-number-mode ":%c")))))
#+end_src
*** directory display
#+begin_src emacs-lisp
(defun shorten-directory (dir max-length)
  "Show up to `max-length' characters of a directory name `dir'."
  (let ((path (reverse (split-string (abbreviate-file-name dir) "/")))
        (output ""))
    (when (and path (equal "" (car path)))
      (setq path (cdr path)))
    (while (and path (< (length output) (- max-length 4)))
      (setq output (concat (car path) "/" output))
      (setq path (cdr path)))
    (when path
      (setq output (concat ".../" output)))
    output))

(defvar mode-line-directory
  '(:propertize
    (:eval (if (buffer-file-name) (concat " " (shorten-directory default-directory 20)) " "))
    face mode-line-directory)
  "Formats the current directory.")
(put 'mode-line-directory 'risky-local-variable t)
#+end_src

*** hide some minor mode
#+begin_src emacs-lisp
; TODO
#+end_src
*** bind it together
#+begin_src emacs-lisp
(setq-default mode-line-format
              '("%e"
                mode-line-front-space
                mode-line-mule-info ; encoding
                mode-line-client
                mode-line-modified
                ;; mode-line-remote -- no need to indicate this specially
                ;; mode-line-frame-identification -- this is for text-mode emacs only
                " "
                ;mode-line-directory
                mode-line-buffer-identification
                " "
                mode-line-position
                ;;(vc-mode vc-mode)  -- I use magit, not vc-mode
                (flycheck-mode flycheck-mode-line)
                " "
                mode-line-modes
                mode-line-misc-info
                mode-line-end-spaces))
#+end_src
* Global enable mode
#+begin_src emacs-lisp
(global-display-line-numbers-mode)
(global-tree-sitter-mode) ;; only in the mode, won't highlight
(keyfreq-mode 1)
(keyfreq-autosave-mode 1)
;; always enable wakatime
(global-wakatime-mode t)
#+end_src

* Programming
** prog-mode hook
#+begin_src emacs-lisp
(add-hook 'prog-mode-hook
          (lambda ()
            (rainbow-delimiters-mode)
            ;; buffer-face
            ;; set the foreground color
;            (setq buffer-face-mode-face '(:foreground "#85ddff"))
;            (buffer-face-mode)
            ))
#+end_src
** Syntax highlight
*** tree-sitter
enable tree-sitter for needed language
#+begin_src emacs-lisp
(add-hook 'go-mode-hook #'tree-sitter-hl-mode)
#+end_src
** Lsp
I use lsp-mode to highlight the code. use lsp-bridge to provide the complement.
*** lsp-mode
set lsp-idle-delay from 0.5 to 0.25 to make it looks faster
#+begin_src emacs-lisp
(setq lsp-idle-delay 0.25)
(setq lsp-semantic-tokens-honor-refresh-requests t); always auto refresh

; disable the lsp-mode's completion
(setq lsp-completion-provider :none)
#+end_src

*** lsp-bridge
use lsp-bridge to do the completion
#+begin_src emacs-lisp
(require 'yasnippet)
(yas-global-mode 1)

(require 'lsp-bridge)
(setq lsp-bridge-enable-hover-diagnostic t)
(global-lsp-bridge-mode)

; disable the diagnostic
(setq lsp-bridge-enable-diagnostics nil)
#+end_src

**** key bindings
我使用键盘的 =magic-fn= （也就是 capslock 可以作为 FN 键使用，然后把 FN+hjkl bind 了一下，所以不需要配置快捷键，直接可以起到类似 ctrl-j/k 切换选项的效果了）
#+begin_src emacs-lisp
;(evil-define-key 'insert 'lsp-bridge-mode (kbd "C-j") 'acm-select-next)
;(evil-define-key 'insert 'lsp-bridge-mode (kbd "C-k") 'acm-select-prev)
#+end_src
** company mode
#+begin_src emacs-lisp
(after! company
  (setq company-global-modes '(not erc-mode circe-mode message-mode help-mode gud-mode vterm-mode c-mode c++-mode python-mode emacs-lisp-mode verilog-mode org-mode rust-mode))
  )
(global-company-mode)
#+end_src
** C/C++
*** lsp
#+begin_src emacs-lisp
(setq lsp-clients-clangd-args '("-j=8"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=never"
                                "--header-insertion-decorators=0"))

(after! lsp-clangd (set-lsp-priority! 'clangd 2))

(defun c-mode-lsp-setup ()
  (setq tree-sitter-mode nil)
  (setq lsp-semantic-tokens-enable t)
  )

(add-hook 'c-mode-common-hook 'c-mode-lsp-setup)
#+end_src
*** format
#+begin_src emacs-lisp
;; google c style
(add-hook 'c-mode-common-hook 'google-set-c-style)
#+end_src
** Rust
*** lsp
#+begin_src emacs-lisp
(defun rust-mode-lsp-setup ()
  (setq tree-sitter-mode nil)
  (setq lsp-semantic-tokens-enable t)
  (setq lsp-rust-analyzer-server-display-inlay-hints t)
  (highlight-indent-guides-mode -1) ; disable highlight indent in rust mode
  )
;
(add-hook 'rust-mode-hook 'rust-mode-lsp-setup)
#+end_src
** Common Lisp
#+begin_src emacs-lisp
(add-hook! lisp-mode-hook #'(lambda () (lsp-bridge-mode -1)))
#+end_src
* Org-mode
** General
#+begin_src emacs-lisp
(after! org
  ;;; org-journal
  (setq org-journal-file-format "%Y%m%d.org")
  (setq org-journal-file-header ":PROPERTIES:
#+LATEX_COMPILER: xelatex
#+LATEX_CLASS: elegentpaper
#+OPTIONS: prop:t
:END:\n\n")
  ;;; org-agenda
  (setq org-agenda-files (directory-files-recursively "~/org/" "\\.org$"))
  ; open with all title close
  (setq org-startup-folded t)
  (setq org-log-done 'time)
  (setq org-format-latex-options (plist-put org-format-latex-options :scale 2.0))
)
#+end_src
** Storage
*** Directory
#+begin_src emacs-lisp
;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")
(setq org-roam-directory "~/org/roam")
#+end_src
** Latex
#+begin_src emacs-lisp
(after! ox-latex
(setq org-latex-src-block-backend 'minted
      org-latex-pdf-process '("latexmk -xelatex -quiet -shell-escape -f %f"))
(add-to-list 'org-latex-logfiles-extensions "bbl")
(add-to-list 'org-latex-logfiles-extensions "tex")
(setq org-latex-remove-logfiles t) ;; ensure the cleanup
(add-to-list 'org-latex-packages-alist '("newfloat" "minted"))
(add-to-list 'org-latex-classes
             '("elegentpaper"
               "\\documentclass[lang=cn]{elegantpaper}
               [NO-DEFAULT-PACKAGES]
               [PACKAGES]
               [EXTRA]"
               ("\\section{%s}" . "\\section*{%s}")
               ("\\subsection{%s}" . "\\subsection*{%s}")
               ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
               ("\\paragraph{%s}" . "\\paragraph*{%s}")
               ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
)
#+end_src
** For export
#+begin_src emacs-lisp
(setq org-export-preserve-breaks t)
#+end_src
* Key bindings
#+begin_src emacs-lisp
(map! :localleader
      :map org-mode-map
      :prefix "l"
      "p" #'org-paste-image-from-windows)
(map! :after company
      :map company-active-map
      "<tab>" #'company-complete-selection)
#+end_src
* Other
** WSL setup && Chinese input (rime)
+ 添加了粘贴图片的支持
+ 使用 rime 输入法，并且自动在特定情况下使用临时英文输入

#+begin_src emacs-lisp
(when (string-match-p "WSL" (shell-command-to-string "uname -a"))
  ;; use windows browser
  ;; note:
  ;; 1. this could make the org reveal.js generated html failed to find
  ;; picture, as our picture pasting is using a absolute path
  ;; 2. make sure to have a windows-browser in path
  (setq browse-url-browser-function 'browse-url-generic
        browse-url-generic-program "windows-browser")
  ;; use a better chinese input
  (setq default-input-method "rime")
  (setq rime-show-candidate 'posframe)
  (setq rime-user-data-dir "~/.doom.d/rime")
  ;; 在 evil-normal-state 中、在英文字母后面以及代码等情况下中自动使用英文
  (setq rime-disable-predicates
        '(rime-predicate-evil-mode-p
          rime-predicate-after-alphabet-char-p
          rime-predicate-punctuation-after-space-cc-p
          rime-predicate-space-after-cc-p
          rime-predicate-current-uppercase-letter-p
          rime-predicate-after-ascii-char-p
          rime-predicate-prog-in-code-p
          rime-predicate-punctuation-line-begin-p
          ))

  (defun rime-auto-change-punctuation-after-inline-mode ()
    "如果当前处于 rime enable 的状态，并且光标的前两个字符为半角标点符号加空格，则将前两个符号替换为一个对应的全角标点。"
    (if (and (rime-mode) (> (point) 2))
        (let ((punc  (char-before (- (point) 1)))
              (blank (char-before)))
          (if (= blank (string-to-char " "))
              (cond
               ((= punc ?,) (progn (delete-char -2) (insert "，")))
               ((= punc ?.) (progn (delete-char -2) (insert "。")))
               ((= punc ?\;) (progn (delete-char -2) (insert "；")))
               ((= punc ?\?) (progn (delete-char -2) (insert "？")))
               ((= punc ?!) (progn (delete-char -2) (insert "！")))
               ((= punc ?\() (progn (delete-char -2) (insert "（")))
               ((= punc ?\)) (progn (delete-char -2) (insert "）"))))))))

  (add-hook! rime-active-mode #'rime-auto-change-punctuation-after-inline-mode)

  ;; org paste image from windows host
  (setq org-startup-with-inline-images t)
  (defun org-paste-image-from-windows ()
    "Paste an image into a time stamped unique-named file in the ~/.org/picture
  and insert a link to this file"
    (interactive)
    (let* ((target-file
            (concat
             (make-temp-name
              (concat
               "~/.org/picture/"
               (format-time-string "%Y%m%d_%H%M%S_")))
             "\.png"))
           (windows-path
            (wsl-to-windows-path target-file))
           (ps-script
            (concat "(Get-Clipboard -Format image).Save('" windows-path "')")))
      (powershell ps-script)

      (if (file-exists-p target-file)
          (progn (insert (concat "[[" target-file "]]"))
                 (org-display-inline-images)
                 (message (concat "saving to " ps-script "..."))
                 )
        (user-error
         "Pasting the image failed.."))
      ))

  (defun wsl-to-windows-path (path)
    "Conver a wsl unix path to its windows path"
    (substring
     (shell-command-to-string (concat "wslpath -w " path)) 0 -1))

  (defun powershell (script)
    "Execute the given script within a powershell and return its return value.
  Note: pwsh should be a valid command that can start a powershell, for example,
  make a symblic link to powershell.exe to ~/.local/bin/powershell"
    (call-process "powershell" nil nil nil
                  "-noprofile"
                  "-Command" (concat "& {" script "}")))
  )
#+end_src
** Profile startup
在一般情况下，注释下面的代码
#+begin_src emacs-lisp
;; profile startup, comment always!
;(require 'benchmark-init)
;(add-hook 'doom-first-input-hook #'benchmark-init/deactivate)
#+end_src
** EAF
使用 eaf ，主要是为了他的浏览器，不过试用之后，仍然感觉不是很满足个人需求，暂时先不用了
#+begin_src emacs-lisp
;(use-package eaf)
;
;(setq env-proxy-ip-port-list (split-string (substring (getenv "http_proxy") 7) "\\:"))
;(setq eaf-proxy-type "http")
;(setq eaf-proxy-host (car env-proxy-ip-port-list))
;(setq eaf-proxy-port (car (cdr env-proxy-ip-port-list)))
;
;(require 'eaf-system-monitor)
;(require 'eaf-image-viewer)
;(require 'eaf-markdown-previewer)
;(require 'eaf-browser)
;(require 'eaf-rss-reader)
;(require 'eaf-pdf-viewer)
#+end_src


config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name "chuj"
      user-mail-address "1768485949@qq.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-dark+
      doom-dark+-blue-modeline t
      doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 20
                           :weight 'normal)
      doom-variable-pitch-font (font-spec :family "DejaVu Sans" :size 20))
(set-fontset-font "fontset-default" 'han '("LiHeiPro" . "ttf"))

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/")


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

;;; global enable mode
(global-display-line-numbers-mode)
(global-tree-sitter-mode) ;; only in the mode, won't highlight
(keyfreq-mode 1)
(keyfreq-autosave-mode 1)
;; always enable wakatime
(global-wakatime-mode t)

(setq server-window 'pop-to-buffer)

;;; hooks
;; tree-sitter
(add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode)
(setq lsp-clients-clangd-args '("-j=8"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=never"
                                "--header-insertion-decorators=0"))
(after! lsp-clangd (set-lsp-priority! 'clangd 2))

(add-to-list 'default-frame-alist '(fullscreen . maximized))
;; we don't want a ruler in shell right? so only in prog mode
(add-hook 'prog-mode-hook #'display-fill-column-indicator-mode)
;; i want to see if there is a tab or whitespace in prog mode
(add-hook 'prog-mode-hook
          (lambda ()
            (whitespace-mode t)
            (rainbow-delimiters-mode)
            ;; set the foreground color
            (setq buffer-face-mode-face '(:foreground "DeepSkyBlue"))
            (buffer-face-mode)
            ))

;; google c style
(add-hook 'c-mode-common-hook 'google-set-c-style)
;; better indent guide
(add-hook 'prog-mode-hook 'highlight-indent-guides-mode)
(setq-default highlight-indent-guides-method 'character)
;; profile startup, comment always!
;;(require 'benchmark-init)
;;(add-hook 'doom-first-input-hook #'benchmark-init/deactivate)

;;; other (i'm consider to name the other)

;; 80 ruler
(setq display-fill-column-indicator-column 80)

;; don't use tab
(setq-default indent-tabs-mode nil)
(setq-default tab-width 2)

(setq keyfreq-excluded-commands
      '(self-insert-command
        forward-char
        backward-char
        previous-line
        next-line
        evil-previous-line
        evil-next-line
        evil-backward-char
        evil-forward-char
        evil-force-normal-state
        evil-insert
        term-send-raw
        org-self-insert-command
        vterm--self-insert
        evil-normal-state
        exit-minibuffer))

;;; key bindings
(map! :localleader
      :map org-mode-map
      :prefix "l"
      "p" #'org-paste-image-from-windows)

;;;#############################################################################
;;; latex
;;;#############################################################################
(after! ox-latex
(setq org-latex-src-block-backend 'minted
      org-latex-pdf-process '("latexmk -xelatex -quiet -shell-escape -f %f"))
(add-to-list 'org-latex-logfiles-extensions "bbl")
(add-to-list 'org-latex-logfiles-extensions "tex")
(setq org-latex-remove-logfiles t) ;; ensure the cleanup
(add-to-list 'org-latex-packages-alist '("newfloat" "minted"))
(add-to-list 'org-latex-classes
             '("elegentpaper"
               "\\documentclass[lang=cn]{elegantpaper}
               [NO-DEFAULT-PACKAGES]
               [PACKAGES]
               [EXTRA]"
               ("\\section{%s}" . "\\section*{%s}")
               ("\\subsection{%s}" . "\\subsection*{%s}")
               ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
               ("\\paragraph{%s}" . "\\paragraph*{%s}")
               ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))
)

;;;#############################################################################
;;; org
;;;#############################################################################
(setq org-roam-directory "~/org/roam")
(after! org
;;; org-journal
(setq org-journal-file-format "%Y%m%d.org")
(setq org-journal-file-header "#+LATEX_COMPILER: xelatex
#+LATEX_CLASS: elegentpaper
#+OPTIONS: prop:t\n\n")
;;; org-agenda
(setq org-agenda-files (directory-files-recursively "~/org/" "\\.org$"))
; open with all title close
(setq org-startup-folded t)
(setq org-log-done 'time)
(setq org-format-latex-options (plist-put org-format-latex-options :scale 2.0))
)

;;;#############################################################################
;;; only in wsl
;;;#############################################################################
(when (string-match-p "WSL" (shell-command-to-string "uname -a"))
  ;; use windows browser
  ;; note:
  ;; 1. this could make the org reveal.js generated html failed to find
  ;; picture, as our picture pasting is using a absolute path
  ;; 2. make sure to have a windows-browser in path
  (setq browse-url-browser-function 'browse-url-generic
        browse-url-generic-program "windows-browser")
  ;; use a better chinese input
  (setq default-input-method "rime")
  (setq rime-show-candidate 'posframe)
  (setq rime-user-data-dir "~/.doom.d/rime")

  ;; org paste image from windows host
  (setq org-startup-with-inline-images t)
  (defun org-paste-image-from-windows ()
    "Paste an image into a time stamped unique-named file in the ~/.org/picture
  and insert a link to this file"
    (interactive)
    (let* ((target-file
            (concat
             (make-temp-name
              (concat
               "~/.org/picture/"
               (format-time-string "%Y%m%d_%H%M%S_")))
             "\.png"))
           (windows-path
            (wsl-to-windows-path target-file))
           (ps-script
            (concat "(Get-Clipboard -Format image).Save('" windows-path "')")))
           (powershell ps-script)

           (if (file-exists-p target-file)
               (progn (insert (concat "[[" target-file "]]"))
                      (org-display-inline-images)
                      (message (concat "saving to " ps-script "..."))
                      )
             (user-error
              "Pasting the image failed.."))
           ))

  (defun wsl-to-windows-path (path)
    "Conver a wsl unix path to its windows path"
    (substring
     (shell-command-to-string (concat "wslpath -w " path)) 0 -1))

  (defun powershell (script)
    "Execute the given script within a powershell and return its return value.
  Note: pwsh should be a valid command that can start a powershell, for example,
  make a symblic link to powershell.exe to ~/.local/bin/powershell"
    (call-process "powershell" nil nil nil
                  "-noprofile"
                  "-Command" (concat "& {" script "}")))
)

;;; custom set face
(custom-set-faces
  '(rainbow-delimiters-depth-1-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "gold"))))
  '(rainbow-delimiters-depth-2-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "LightGreen"))))
  '(rainbow-delimiters-depth-3-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "magenta1"))))
  '(rainbow-delimiters-depth-4-face
    ((t (:inherit rainbow-delimiters-base-face :foreground "beige")))))
